pytorch_lightning/callbacks/model_checkpoint.py

```diff
@@ -166,17 +174,21 @@ def __init__(
         save_last: Optional[bool] = None,
         save_top_k: Optional[int] = None,
         save_weights_only: bool = False,
+        every_n_epochs: int = 1,
+        every_n_batches: int = -1,
```

Comment:
for `every_n_batches`? would we want `every_n_epochs` to be optional too?

one choice is to make both `None` by default, and then in the constructor, if neither are set, then we default to `every_n_epochs=1` so it runs after each validation epoch. 

And if someone wants to checkpoint after batches only, they can do `ModelCheckpoint(..., every_n_batches=N)` without needing to toggle off the `every_n_epochs` flag. What do you think?

Reply:
Yes, both optional. The question is whether to assume they are exclusive and we go with

> make both None by default, and then in the constructor, if neither are set, then we default to every_n_epochs=1 so it runs after each validation epoch.

Or not, have `every_n_epochs=1` by default and have the user toggle off if necessary.

Does allowing having both active make sense to you?

Reply:
I think it's unlikely that both will be used together, but they aren't mutually exclusive. if you want to checkpoint every N batches, you can. if you want to checkpoint every M epochs, you can. if you set both, we'll do both 

Since this regular `int` is more restrictive than `Optional[int]` wdyt about going with this approach for now, and if we get feedback that this is confusing, then we can make them both optional to maintain backwards compatibility?

Reply:
Sounds good to me. To recap:

- `-1`: always
- `0`: disabled
- `n`: every n

both can be set at the same time

defaults:
- `every_n_epochs=1`
- `every_n_steps=0`

Correct?

Reply:
correct!

Reply:
@carmocca @awaelchli coming back to this, after looking at the complexity in supporting `monitor` and `save_top_k`, it makes more sense to keep them as mutually exclusive. 

- So we can make both `every_n_train_steps` and `every_n_val_epochs` as `Optional[int]`
- users toggle on whichever they're going to use
- we raise an exception if both are set and positive

Reply:
That's fine with me, especially if we support many callbacks